fdb-record-layer-lucene/src/main/java/com/apple/foundationdb/record/lucene/codec/PrefetchableBufferedChecksumIndexInput.java

```diff
@@ -0,0 +1,104 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.apple.foundationdb.record.lucene.codec;
+
+import com.apple.foundationdb.record.lucene.directory.FDBIndexInput;
+import org.apache.lucene.store.BufferedChecksumIndexInput;
+import org.apache.lucene.store.IndexInput;
+
+import java.io.IOException;
+
+/**
+ * An Index Input that attempts to keep a 10 block buffer in front or at the read at a minimum.  The goal for this
+ * class is to provide some parallelism for fetches from FDB when we are attempting to merge segments.  The signaling
+ * of when to prefetch occurs in the FDBIndexOutput#copyBytes method.
+ *
+ */
+public class PrefetchableBufferedChecksumIndexInput extends BufferedChecksumIndexInput {
+    private final IndexInput indexInput;
+    private static final int PREFETCH_BLOCKS = 10; // Attempt to keep a large enough buffer
+    /**
+     * GreatestFetchBlock from the underlying FDBIndexInput, -1 if not fetched.
+     *
+     */
+    private int greatestFetchedBlock;
+    /**
+     * Maximum fetched block based on the numberOfBytesToPrefetch.
+     *
+     */
+    private int maximumFetchedBlock;
+    /**
+     * Checks whether it is an instance of FDBIndexInput.
+     *
+     */
+    private final boolean supportsPrefetch;
+
+    public PrefetchableBufferedChecksumIndexInput(final IndexInput indexInput) {
+        super(indexInput);
+        this.indexInput = indexInput;
+        this.supportsPrefetch =  indexInput instanceof FDBIndexInput;
+    }
+
+    /**
+     * This method sets the total number of bytes to expect and then attempts to read ahead at least 10 blocks or the
+     * remaining number of blocks while reading.
+     *
+     * @param numberOfBytesToPrefetch providing a signal to the total number of bytes to prefetch
+     */
+    public void setExpectedBytes(long numberOfBytesToPrefetch) {
+        if (supportsPrefetch) {
+            this.maximumFetchedBlock = ((FDBIndexInput) indexInput).getBlock(numberOfBytesToPrefetch - 1);
+            this.greatestFetchedBlock = ((FDBIndexInput) indexInput).getBlock(0L) - 1;
+            checkBuffer(0L);
+        }
+    }
+
+    /**
+     * Based on the greatestFetchBlock and the current positon of the underlying FDBIndexInput, it will decide how
+     * many blocks to fetch.
+     *
+     * @param length position
+     */
+    public void checkBuffer(long length) {
+        if (supportsPrefetch) {
+            int blocksToBuffer = ((FDBIndexInput) indexInput).getBlock(indexInput.getFilePointer() + length) +
+                                 Math.min(PREFETCH_BLOCKS, maximumFetchedBlock - greatestFetchedBlock) -
+                                 greatestFetchedBlock - 1;
```

Comment:
I think the expectation of such a windowing algorithm is that on the first call it will fetch the window size number of blocks and then each time the advancing cursor crosses into a new block it will read another so long as that would be within the target range.

This calculation does something slightly different. Consequently, it slows down to fetching a new block every other block as it nears the end. The issue with which is that it gets increasingly likely that something will not be prefetched and the very last block will not be prefetched at all.

Reply:
For example, consider wanting to copy 25 blocks and note how each call to `checkBuffer` will work as the block advances:

|block|maximum|greatest|new max|new gt|
|----:|------:|-------:|------:|-----:|
|0|24|-1|24|9|
|0|24|9|24|9|
|1|24|9|24|10|
|1|24|10|24|10|
|2|24|10|24|11|
|2|24|11|24|11|
|3|24|11|24|12|
|3|24|12|24|12|
|4|24|12|24|13|
|4|24|13|24|13|
|5|24|13|24|14|
|5|24|14|24|14|
|6|24|14|24|15|
|6|24|15|24|15|
|7|24|15|24|15|
|8|24|15|24|16|
|8|24|16|24|16|
|9|24|16|24|16|
|10|24|16|24|17|
|10|24|17|24|17|
|11|24|17|24|17|
|12|24|17|24|18|
|12|24|18|24|18|
|13|24|18|24|18|
|14|24|18|24|19|
|14|24|19|24|19|
|15|24|19|24|19|
|16|24|19|24|20|
|16|24|20|24|20|
|17|24|20|24|20|
|18|24|20|24|21|
|18|24|21|24|21|
|19|24|21|24|21|
|20|24|21|24|22|
|20|24|22|24|22|
|21|24|22|24|22|
|22|24|22|24|23|
|22|24|23|24|23|
|23|24|23|24|23|
|24|24|23|24|24|
|24|24|24|24|24|

Reply:
Fixed this, good catch.